File: src/test/java/leets/leenk/domain/feed/domain/service/ReactionSaveServiceTest.java
package leets.leenk.domain.feed.domain.service;

import leets.leenk.domain.feed.domain.entity.Feed;
import leets.leenk.domain.feed.domain.entity.Reaction;
import leets.leenk.domain.feed.domain.repository.ReactionRepository;
import leets.leenk.domain.feed.test.FeedTestFixture;
import leets.leenk.domain.feed.test.ReactionTestFixture;
import leets.leenk.domain.feed.test.UserTestFixture;
import leets.leenk.domain.user.domain.entity.User;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
public class ReactionSaveServiceTest {
    @Mock
    private ReactionRepository reactionRepository;

    @InjectMocks
    private ReactionSaveService reactionSaveService;

    @Test
    @DisplayName("reactionSave 테스트")
    void reactionSave() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        Reaction reaction = ReactionTestFixture.createReaction(feed, user, 1L);

        given(reactionRepository.save(reaction)).willReturn(reaction);

        // when
        Reaction result = reactionSaveService.save(reaction);

        // then
        assertThat(result).isSameAs(reaction);
        then(reactionRepository).should().save(reaction);
    }
}

File: src/test/kotlin/leets/leenk/domain/feed/domain/service/ReactionSaveServiceTest.kt
package leets.leenk.domain.feed.domain.service

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import leets.leenk.domain.feed.domain.repository.ReactionRepository
import leets.leenk.domain.feed.test.FeedTestFixture
import leets.leenk.domain.feed.test.ReactionTestFixture
import leets.leenk.domain.feed.test.UserTestFixture

class ReactionSaveServiceTest :
    BehaviorSpec({
        val reactionRepository = mockk<ReactionRepository>()
        val reactionSaveService = ReactionSaveService(reactionRepository)

        Given("리액션이 주어졌을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val reaction = ReactionTestFixture.createReaction(feed, user, 1L)

            every { reactionRepository.save(reaction) } returns reaction

            When("리액션을 저장하면") {
                val result = reactionSaveService.save(reaction)

                Then("저장된 리액션이 반환되어야 한다") {
                    result shouldBe reaction
                    verify { reactionRepository.save(reaction) }
                }
            }
        }
    })
